annotation/tests/test_tasks_services.py

```diff
@@ -0,0 +1,322 @@
+        pytest.fail("FieldConstraintError was raised unexpectedly")
+
+
+@patch("sqlalchemy.orm.Session", spec=True)
+def test_check_cross_annotating_pages_raise_error_when_page_already_annotated(
+    mock_session: Mock,
+):
+    db_session = mock_session()
+    task_info = {"user_id": 1, "file_id": 2, "job_id": 3, "pages": {4, 5}}
+    existing_pages = [4, 5]
+
+    mock_query = db_session.query.return_value
+    mock_query.filter.return_value.all.return_value = [(existing_pages,)]
+
+    expected_error_message = (
+        "within cross validation job user can't validate file's pages "
+        "that are already distributed in annotation tasks for this user: "
+        "{4, 5}"
+    )
+
+    with pytest.raises(FieldConstraintError, match=expected_error_message):
+        check_cross_annotating_pages(db_session, task_info)
+
+
+@pytest.mark.parametrize(
+    (
+        "is_validation",
+        "failed",
+        "annotated",
+        "not_processed",
+        "annotation_user",
+        "validation_user",
+        "expected_error_message",
+    ),
+    (
+        (
+            True,
+            {1},
+            set(),
+            set(),
+            False,
+            True,
+            "Missing `annotation_user_for_failed_pages` param "
+            "for failed pages.",
+        ),
+        (
+            True,
+            set(),
+            {2},
+            set(),
+            True,
+            False,
+            "Missing `validation_user_for_reannotated_pages` param "
+            "for edited pages.",
+        ),
+        (
+            True,
+            set(),
+            set(),
+            set(),
+            True,
+            True,
+            "Validator did not mark any pages as failed, thus "
+            "`annotation_user_for_failed_pages` param should be null.",
+        ),
+        (
+            True,
+            {1},
+            set(),
+            set(),
+            True,
+            True,
+            "Validator did not edit any pages, thus "
+            "`validation_user_for_reannotated_pages` param should be null.",
+        ),
+        (
+            True,
+            {1},
+            {2},
+            {3},
+            True,
+            True,
+            "Cannot finish validation task. There are not processed pages: "
+            "{3}",
+        ),
+    ),
+)
+def test_validate_user_actions_raises_error_for_invalid_user_states(
+    is_validation: bool,
+    failed: set,
+    annotated: set,
+    not_processed: set,
+    annotation_user: bool,
+    validation_user: bool,
+    expected_error_message: str,
+):
+    with pytest.raises(HTTPException) as excinfo:
+        validate_user_actions(
+            is_validation=is_validation,
+            failed=failed,
+            annotated=annotated,
+            not_processed=not_processed,
+            annotation_user=annotation_user,
+            validation_user=validation_user,
+        )
+    assert excinfo.value.detail == expected_error_message
```

Comment:
https://docs.pytest.org/en/stable/how-to/assert.html#matching-exception-messages

Reply:
Ok.

Reply:
I attempted to use match within pytest.raises for testing error messages, but since the error description is within the detail attribute of the exception, I think it's more accurate to directly assert excinfo.value.detail against the expected message.

Reply:
I did not get the problem. Could you provide screenshot/examples why match wouldn't work here?

Reply:
I reviewed the issue and found that the reason for the AssertionError with the message Regex pattern '.*Cannot finish validation task. There are not processed pages.*' does not match '' is that the error message is stored in the detail attribute of the HTTPException.
The match parameter in pytest.raises does not compare the detail attribute directly; instead, it attempts to match the entire exception message. Since the detail field is where the message is stored, match ends up comparing against an empty string, which leads to the failure.